From e0115b2d28e1a030c6158f2c57a2ebfd62507d1b Mon Sep 17 00:00:00 2001
From: Christopher Albert <albert@tugraz.at>
Date: Sun, 22 Mar 2026 06:53:08 +0100
Subject: [PATCH] fortran: Avoid rollback ICE after invalid SELECT TYPE

Distinguish MATCH_NO from MATCH_ERROR in gfc_match_select_type so a
failed expression match does not free a partially broken temporary
namespace through the normal rollback path.

PR fortran/84245
gcc/fortran/ChangeLog:

	* match.cc (gfc_match_select_type): Free the temporary namespace
	only on MATCH_NO. Return immediately on MATCH_ERROR.

gcc/testsuite/ChangeLog:

	* gfortran.dg/pr84245.f90: New test.

Signed-off-by: Christopher Albert <albert@tugraz.at>
---
 gcc/fortran/match.cc                  | 7 ++++++-
 gcc/testsuite/gfortran.dg/pr84245.f90 | 7 +++++++
 2 files changed, 13 insertions(+), 1 deletion(-)
 create mode 100644 gcc/testsuite/gfortran.dg/pr84245.f90

diff --git a/gcc/fortran/match.cc b/gcc/fortran/match.cc
index b37337d505f..62bdb8687e5 100644
--- a/gcc/fortran/match.cc
+++ b/gcc/fortran/match.cc
@@ -7510,12 +7510,17 @@ gfc_match_select_type (void)
   else
     {
       m = gfc_match (" %e ", &expr1);
-      if (m != MATCH_YES)
+      if (m == MATCH_NO)
 	{
 	  std::swap (ns, gfc_current_ns);
 	  gfc_free_namespace (ns);
 	  return m;
 	}
+      /* On MATCH_ERROR, the temporary block namespace may already contain
+	 broken state from the failed expression match.  Avoid freeing it
+	 through the normal rollback path.  */
+      else if (m == MATCH_ERROR)
+	return m;
     }
 
   m = gfc_match (" )%t");
diff --git a/gcc/testsuite/gfortran.dg/pr84245.f90 b/gcc/testsuite/gfortran.dg/pr84245.f90
new file mode 100644
index 00000000000..f43541fd4b2
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/pr84245.f90
@@ -0,0 +1,7 @@
+! { dg-do compile }
+! { dg-options "-w" }
+
+program p
+  select type (x%a) ! { dg-error "Unexpected '%'.*nonderived-type variable 'x'" }
+  end select        ! { dg-error "Expecting END PROGRAM statement" }
+end
-- 
2.53.0

